src/tools/add_entity.cc

```diff
@@ -100,39 +100,44 @@ struct add_entity_tool : tool {
         ch->entities.push_back(e);
     }
 
+    // press to rotate once, hold to rotate continuously
     void alt_use(raycast_info *rc) override {
-        do {
-            entity_name_index++;
-            if (entity_name_index >= entity_names.size()) {
-                entity_name_index = 0;
-            }
-        } while (!entity_stubs[entity_names[entity_name_index]].get_component<placeable_component_stub>());
-
         auto rotate = get_rotate();
-        if (rotate == 0) {
-            cur_rotate = 0;
-        }
-        else {
-            cur_rotate += rotate / 2;
-            cur_rotate = cur_rotate - (cur_rotate % rotate);
-        }
+
+        cur_rotate -= rotate;
+        cur_rotate %= 360;
     }
 
     void long_use(raycast_info *rc) override {
     }
 
+    // hold to rotate continuously, press to rotate once
     void long_alt_use(raycast_info *rc) override {
         auto rotate = get_rotate();
 
         if (frame_info.elapsed >= last_rotate_time + 1.0 / rotate_tick_rate) {
-            cur_rotate += rotate;
+            cur_rotate -= rotate;
             cur_rotate %= 360;
             last_rotate_time = frame_info.elapsed;
-            printf("%d\n", cur_rotate);
         }
     }
 
     void cycle_mode() override {
+        do {
+            entity_name_index++;
+            if (entity_name_index >= entity_names.size()) {
+                entity_name_index = 0;
+            }
+        } while (!entity_stubs[entity_names[entity_name_index]].get_component<placeable_component_stub>());
+
+        auto rotate = get_rotate();
+        if (rotate == 0) {
+            cur_rotate = 0;
+        }
+        else {
+            cur_rotate += rotate / 2;
+            cur_rotate = cur_rotate - (cur_rotate % rotate);
```

Comment:
Hidden assumption here about rotate>0 ?

Reply:
yes

Reply:
Rather, about cur_rotate>0. This might be OK.